File: apps/mobile/test/core/city_progress_test.dart
/// ⭐⭐ ŞEHİR SAYAÇLARI — web ile AYNI dosyadan okunan vektörler.
///
/// `clock_test.dart` ile aynı sözleşme: `packages/contracts/fixtures/city-progress-vectors.json`
/// iki uygulamayı da ölçüyor, dosya kopyalanmıyor.
///
/// ⚠️ Ölçülen iki hesap da web'de **canlı hatalardan** doğdu: üretim bandı 2026-07-28'de
/// kullanıcının bildirdiği donma, kaynak sayacı ise yoklama aralığı 60 sn'ye çıkınca görünür
/// olan sıçrama. İkisinin de web'de testi yoktu; mobil port ikisini de sınanabilir yaptı.
library;

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:mobilwar/core/city_progress.dart';

/// ⚠️ `flutter test` çalışma dizini paket kökü (`apps/mobile`).
final _vectors =
    jsonDecode(
          File(
            '../../packages/contracts/fixtures/city-progress-vectors.json',
          ).readAsStringSync(),
        )
        as Map<String, dynamic>;

List<Map<String, dynamic>> _grup(String ad) =>
    (_vectors[ad] as List<dynamic>).cast<Map<String, dynamic>>();

int _ms(String iso) => DateTime.parse(iso).millisecondsSinceEpoch;

void main() {
  group('⭐⭐ ortak vektörler — üretim bandı', () {
    for (final v in _grup('unitProgress')) {
      test(v['ad'] as String, () {
        final got = unitProgress(
          ProgressInput(
            startedAt: v['startedAt'] as String,
            count: (v['count'] as num?)?.toInt(),
            perUnitSeconds: v['perUnitSeconds'] as num?,
          ),
          _ms(v['now'] as String),
        );

        final beklenen = v['beklenen'] as Map<String, dynamic>?;
        if (beklenen == null) {
          expect(got, isNull);
          return;
        }
        expect(got, isNotNull);
        expect(got!.produced, beklenen['produced']);
        expect(got.remaining, beklenen['remaining']);
        expect(got.finished, beklenen['finished']);
        expect(got.unitStart, _ms(beklenen['unitStart'] as String));
        expect(got.unitEnd, _ms(beklenen['unitEnd'] as String));
      });
    }
  });

  group('⭐⭐ ortak vektörler — kaynak sayacı', () {
    for (final v in _grup('resources')) {
      test(v['ad'] as String, () {
        final got = extrapolateResources(
          ResourceInput(
            gold: v['gold'] as num,
            food: v['food'] as num,
            goldPerHour: v['goldPerHour'] as num,
            foodPerHour: v['foodPerHour'] as num,
            serverNow: v['serverNow'] as String,
          ),
          _ms(v['now'] as String),
        );

        final beklenen = v['beklenen'] as Map<String, dynamic>;
        // ⚠️ TAM eşitlik — `closeTo` DEĞİL. İki dil de IEEE754 double kullanıyor ve işlem
        // sırası aynı; yaklaşık karşılaştırma tam olarak yakalamak istediğimiz farkı gizlerdi.
        expect(got.gold, beklenen['gold']);
        expect(got.food, beklenen['food']);
      });
    }
  });

  group('sözleşme bütünlüğü', () {
    test('⚠️ vektör dosyası gerçekten okundu (sessiz boş küme değil)', () {
      expect(_grup('unitProgress').length, greaterThan(8));
      expect(_grup('resources').length, greaterThan(5));
    });

    test('⭐ çıpa YALNIZ `startedAt` — aynı girdiye aynı cevap', () {
      const q = ProgressInput(
        startedAt: '2026-08-15T12:00:00.000Z',
        count: 10,
        perUnitSeconds: 60,
      );
      final now = _ms('2026-08-15T12:05:00.000Z');
      expect(unitProgress(q, now)!.produced, 5);
    });
  });
}

File: apps/web/test/city-progress.test.ts
/**
 * ⭐⭐ ŞEHİR SAYAÇLARI — üretim bandı ve kaynak ekstrapolasyonu.
 *
 * **Neden şimdi yazıldı:** iki hesabın da bugüne kadar HİÇ testi yoktu, çünkü ikisi de bir JSX
 * bileşeninin içine gömülüydü (`City.tsx`, `Shell.tsx`). Mobil port ikisini de
 * `lib/city-progress.ts`e çıkardı ve ilk kez sınanabilir hâle getirdi — oysa `unitProgress`
 * **canlıda kullanıcının bildirdiği bir hatanın** düzeltmesiydi (2026-07-28) ve o düzeltmeyi
 * koruyan hiçbir şey yoktu.
 *
 * ⭐⭐ Vektörler `packages/contracts/fixtures/city-progress-vectors.json`ta ve Dart testi
 * **aynı dosyayı** okuyor: iki istemcinin bu sayaçlarda ayrışması, aynı şehri iki cihazdan
 * açan oyuncuya iki farklı asker sayısı göstermek olurdu.
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { extrapolateResources, unitProgress } from '../src/lib/city-progress.ts';

const VECTORS = JSON.parse(
  readFileSync(
    fileURLToPath(new URL('../../../packages/contracts/fixtures/city-progress-vectors.json', import.meta.url)),
    'utf8',
  ),
) as {
  unitProgress: {
    ad: string; startedAt: string; now: string; count: number | null;
    perUnitSeconds: number | null;
    beklenen: null | {
      produced: number; remaining: number; finished: boolean;
      unitStart: string; unitEnd: string;
    };
  }[];
  resources: {
    ad: string; gold: number; food: number; goldPerHour: number; foodPerHour: number;
    serverNow: string; now: string; beklenen: { gold: number; food: number };
  }[];
};

describe('⭐⭐ ortak vektörler — üretim bandı', () => {
  it.each(VECTORS.unitProgress)('$ad', (v) => {
    const got = unitProgress(
      { startedAt: v.startedAt, count: v.count, perUnitSeconds: v.perUnitSeconds },
      Date.parse(v.now),
    );

    if (v.beklenen === null) {
      expect(got).toBeNull();
      return;
    }
    expect(got).not.toBeNull();
    expect(got!.produced).toBe(v.beklenen.produced);
    expect(got!.remaining).toBe(v.beklenen.remaining);
    expect(got!.finished).toBe(v.beklenen.finished);
    expect(got!.unitStart).toBe(Date.parse(v.beklenen.unitStart));
    expect(got!.unitEnd).toBe(Date.parse(v.beklenen.unitEnd));
  });
});

describe('⭐⭐ ortak vektörler — kaynak sayacı', () => {
  it.each(VECTORS.resources)('$ad', (v) => {
    const got = extrapolateResources(
      {
        gold: v.gold, food: v.food,
        goldPerHour: v.goldPerHour, foodPerHour: v.foodPerHour,
        serverNow: v.serverNow,
      },
      Date.parse(v.now),
    );
    // ⚠️ `toBe` — `toBeCloseTo` DEĞİL. İki dil de IEEE754 double kullanıyor ve işlem sırası
    // aynı; yaklaşık karşılaştırma tam olarak yakalamak istediğimiz farkı gizlerdi.
    expect(got.gold).toBe(v.beklenen.gold);
    expect(got.food).toBe(v.beklenen.food);
  });
});

describe('sözleşme bütünlüğü', () => {
  it('⚠️ vektör dosyası gerçekten okundu (sessiz boş küme değil)', () => {
    expect(VECTORS.unitProgress.length).toBeGreaterThan(8);
    expect(VECTORS.resources.length).toBeGreaterThan(5);
  });

  /**
   * ⚠️ Sunucunun `done`/`remaining` alanları bilerek KULLANILMIYOR (bayat). Bu test o kararı
   * kilitliyor: girdi tipinde o alanlar hiç yok, yani biri onları okumaya kalksa derlenmez.
   * Burada ölçülen şey ise hesabın gerçekten yalnız `startedAt`e dayandığı.
   */
  it('⭐ çıpa YALNIZ `startedAt` — aynı girdiye aynı cevap, geçmişten bağımsız', () => {
    const q = { startedAt: '2026-08-15T12:00:00.000Z', count: 10, perUnitSeconds: 60 };
    const now = Date.parse('2026-08-15T12:05:00.000Z');
    expect(unitProgress(q, now)).toEqual(unitProgress(q, now));
    expect(unitProgress(q, now)!.produced).toBe(5);
  });
});
